README.md
# oozie-tools

Collection of tools to assit with [Oozie](http://oozie.apache.org/) workflow creation, submission, and reporting.

## Belt

Intelligent Oozie belt-fed job submission based on HDFS directory path input. Recursively searches for workflow XMLs and automagically generates companion job.properties to accompany workflow submission. Uses HDFS and Yarn HA values when detected.

##### Prerequisites

- User must have Oozie ProxyUser permissions in order to submit job
- Must be using [kitesdk](http://kitesdk.org/docs/current/) >= `1.1.0` libraries to utilize Hadoop HA
  - kite-data-core-1.1.0.jar
  - kite-data-hive-1.1.0.jar
  - kite-data-mapreduce-1.1.0.jar
  - kite-hadoop-compatibility-1.1.0.jar

##### Setup

```bash
# clone the repo
$ git clone https://github.com/hartrh/oozie-tools.git
```

Using the `templates/config_example.txt` config as an example, create a text file within the `config/` directory using the naming convention `provider_database_classification_catalog.txt`, with the appropriate values. All slashes (`/`) must be escaped (`\`).

```
# credentials and connection string for example.com
sql_user: imauser
sql_pass: imapassword
conn_string: jdbc:jtds:sqlserver:\/\/example.service._color.consul:1453\/catalog;domain=EXAMPLE.COM;useCursors=true
```

##### Usage

```bash
# execute belt
$ oozie-tools/belt.sh -w 'hdfs/path/to/recursively/search/for/workflows'
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased](unreleased)

- no new features in development at this time

## [0.3.2](https://github.com/hartrh/oozie-tools/compare/0.3.2...0.3.1) (Mar 6, 2017)

BUG FIXES:

- exit script when last table is processed instaed of falsifying running boolean

## [0.3.1](https://github.com/hartrh/oozie-tools/compare/0.3.1...0.3.0) (Mar 6, 2017)

BUG FIXES:

- fix table range argument assignment
- add log statement for job submittion

## [0.3.0](https://github.com/hartrh/oozie-tools/compare/0.3.0...0.2.1) (Mar 6, 2017)

FEATURES

- add flags for start_table and end_table to allow submission of table range.

## [0.2.1](https://github.com/hartrh/oozie-tools/compare/0.2.1...0.2.0) (Feb 15, 2017)

BUG FIXES:

- fix config parsing

## [0.2.0](https://github.com/hartrh/oozie-tools/compare/0.2.0...0.1.2) (Feb 15, 2017)

FEATURES:

- modify the way configurations are consumed.

BUG FIXES:

- previous versions did not support multiple belt jobs running concurrently 

## [0.1.2](https://github.com/hartrh/oozie-tools/compare/0.1.2...0.1.1) (Feb 15, 2017)

BUG FIXES:

- added 'sudo' to environmnet color detection command

## [0.1.1](https://github.com/hartrh/oozie-tools/compare/0.1.1...0.1.0) (Feb 15, 2017)

BUG FIXES:

- fix branching issue

## [0.1.0](https://github.com/hartrh/oozie-tools/compare/0.1.0...0.1.0) (Feb 15, 2017)

FEATURES:

- initial commit


floozie.sh
#!/usr/bin/env bash

#------------------------------------------------------------------------------
# FUNCTIONS
#------------------------------------------------------------------------------

# function: show_help
function show_help () {
  cat << EOF
  Usage: $0 -w 'workflow_dir'

    -w workflow_dir     directory containing oozie workflows for processing.
                        directory provided should be in th«e following format:
                          provider/database/classification/catalog/schema
                        provider, database, classification, catalog and schema
                          variables are reversed from workflow_dir path.
    -h help             help documentation

EOF
}

function is_workflow () {
  if [[ ${obj} =~ (.*)\/workflow.xml$ ]]; then
    return 0;
  else
    return 1;
  fi
}

#------------------------------------------------------------------------------
# ARGUEMENTS
#------------------------------------------------------------------------------

while getopts :n:w:h FLAG; do
  case $FLAG in
    w)  #set option "w"
      workflow_dir=$OPTARG;
      ;;
    h|help)  #show help
      show_help;
      exit 0;
      ;;
  esac
done

shift $((OPTIND-1))  #This tells getopts to move on to the next argument.

# verify arguements have been set
if [ -z ${workflow_dir} ]; then
  echo "==> ERROR: Please specify the 'workflow_dir' arguement. Aborting script.";
  echo "==> Run '$0 -help' for additional info.";
  exit 1;
fi

#------------------------------------------------------------------------------
# VARIABLES
#------------------------------------------------------------------------------

SCRIPTPATH=`dirname "${BASH_SOURCE[0]}"`;
configs_dir="${SCRIPTPATH}/configs";
properties_dir="${SCRIPTPATH}/properties";
reports_dir="${SCRIPTPATH}/reports";
templates_dir="${SCRIPTPATH}/templates"
color="teal"


#------------------------------------------------------------------------------
# MAIN
#------------------------------------------------------------------------------

# get hadoop cluster info
namenode=`hdfs getconf -confkey dfs.nameservices`
jobtracker=`hdfs getconf -confKey yarn.resourcemanager.cluster-id`

# get hdfs workflow list
for obj in `grep -P '^(?!(Found\s([0-9]*)\sitems))' <(hadoop fs -ls -R ${workflow_dir}) | awk '{ print $NF }'`; do
  if is_workflow ${obj}; then

    # notify
    echo "==> Workflow file found: ${obj}"

    # set timestamps
    month=`date "+%m"`;
    day=`date "+%d"`;
    year=`date "+%Y"`;

    # reverse vars from obj path
    IFS='/' read -r -a path <<< "${obj}"
    if [ ${#path[@]} -ge 6 ]; then
      if [[ "${path[${#path[@]}-6]}" == "mdr" && "${path[${#path[@]}-5]}" == "standard" ]]; then
        provider="${path[${#path[@]}-7]}";
        database="${path[${#path[@]}-6]}";
        classification="${path[${#path[@]}-5]}";
        catalog="${path[${#path[@]}-4]}";
        schema="${path[${#path[@]}-3]}";
        table="${path[${#path[@]}-2]}";
      else
        provider="${path[${#path[@]}-6]}";
        database="${path[${#path[@]}-5]}";
        classification="all";
        catalog="${path[${#path[@]}-4]}";
        schema="${path[${#path[@]}-3]}";
        table="${path[${#path[@]}-2]}";
      fi
    else
      echo "==> ERROR: ${obj} needs to be a more complete path. Aborting script.";
      echo "==> Run '$0 -help' for additional info.";
      exit 1;
    fi

    # source config
    if [ "${classification}" != "all" ]; then
      config="${configs_dir}/${provider}_${database}_${classification}_${catalog}.sh"
    else
      config="${configs_dir}/${provider}_${database}_${catalog}.sh"
    fi
    if [ -f ${config} ]; then
      source ${config}
    else
      echo "==> ERROR: missing ${config} file. Aborting script.";
      exit 1;
    fi

    # construct job.properties output
    job_properties_dir="${properties_dir}/${provider}/${database}/${classification}/${catalog}/${schema}/${table}"
    job_properties_file="${job_properties_dir}/${table}_${year}${month}${day}_job.properties"

    # create job.properties file if missing
    if [ ! -f "${job_properties_file}" ]; then
      echo "==> Creating ${job_properties_file}";
      mkdir -p "${job_properties_dir}"
      if [ "${classification}" != "all" ]; then
        cp "${templates_dir}/job.properties" "${job_properties_file}";
      else
        cp "${templates_dir}/job_sans_class.properties" "${job_properties_file}";
      fi 

      # source credentials
      source ${config}

      # update values
      sed -i "s/_namenode/${namenode}/g" ${job_properties_file}
      sed -i "s/_jobtracker/${jobtracker}/g" ${job_properties_file}
      sed -i "s/_month/${month}/g" ${job_properties_file}
      sed -i "s/_day/${day}/g" ${job_properties_file}
      sed -i "s/_year/${year}/g" ${job_properties_file}
      sed -i "s/_provider/${provider}/g" ${job_properties_file}
      sed -i "s/_database/${database}/g" ${job_properties_file}
      sed -i "s/_classification/${classification}/g" ${job_properties_file}
      sed -i "s/_catalog/${catalog}/g" ${job_properties_file}
      sed -i "s/_schema/${schema}/g" ${job_properties_file}
      sed -i "s/_table/${table}/g" ${job_properties_file}
      sed -i "s/_sql_user/${SQL_USER}/g" ${job_properties_file}
      sed -i "s/_sql_pass/${SQL_PASS}/g" ${job_properties_file}
      sed -i "s/_conn_string/${CONN_STRING}/g" ${job_properties_file}
      sed -i "s/_color/${color}/g" ${job_properties_file}
    fi

    # submit workflow to oozie
    job_output=$(oozie job -run -doas hdfs -oozie http://oozie.service.${color}.consul:11000/oozie -config ${job_properties_file});
    job_id=$( echo "${job_output}" | awk '{ print $2 }' );

    # verify job was submitted
    if [ -z "${job_id}" ]; then
      echo "==> ERROR: Job was not sucessfully submitted. Aborting script.";
      exit 1;
    fi

    # wait for job to complete
    while grep -q -P "Status\s+:\s+RUNNING" <(oozie job -info ${job_id} -oozie http://oozie.service.${color}.consul:11000/oozie); do
      sleep 30;
    done

    # query oozie for job metrics
    mapfile -t metrics < <(grep -P "Status\s+:\s+(\w+)|Started\s+\:\s+([a-zA-Z0-9:-\s]+)|Ended\s+\:\s+([a-zA-Z0-9:-\s]+)" <(oozie job -info ${job_id} -oozie http://oozie.service.${color}.consul:11000/oozie))
    status=$(echo "${metrics[0]}" | awk -F': ' '{ print $2 }')
    start_time=$(echo "${metrics[1]}" | awk -F': ' '{ print $2 }')
    end_time=$(echo "${metrics[2]}" | awk -F': ' '{ print $2 }')

    # create report file
    report="${reports_dir}/${provider}_${database}_${classification}_${catalog}_${year}${month}${day}.csv"
    if [ ! -f ${report} ]; then
      if [ -f "${reports_dir}/example.csv" ]; then
        cp "${reports_dir}/example.csv" ${report};
      else
        echo "==> ERROR: Template not found ${reports_dir}/example.csv. Aborting script.";
        exit 1;
      fi
    fi

    # report metrics
    echo "${provider},${database},${classification},${catalog},${schema},${table},${status},${start_time},${end_time},${job_id}" >> ${report}

  fi
done


belt.sh
#!/usr/bin/env bash

#------------------------------------------------------------------------------
# FUNCTIONS
#------------------------------------------------------------------------------

# function: show_help
function show_help () {
  cat << EOF
  Usage: $0 -w 'workflow_dir'

    -w workflow_dir       hdfs directory containing oozie workflows for processing.

                          directory provided should be in the following format:
                            'provider/database/classification/catalog/schema'

    -a first_table        first table name to process if range is provided

    -z last_table         last table name to process if range is provided

    -h help               help documentation

EOF
}

function is_workflow () {
  if [[ ${obj} =~ (.*)\/workflow.xml$ ]]; then
    return 0;
  else
    return 1;
  fi
}

#------------------------------------------------------------------------------
# ARGUEMENTS
#------------------------------------------------------------------------------

while getopts :w:a:z:h FLAG; do
  case $FLAG in
    w)  #set option "w"
      workflow_dir=$OPTARG;
      ;;
    a)  #set option "a"
      first_table=$OPTARG;
      ;;
    z)  #set option "z"
      last_table=$OPTARG;
      ;;
    h|help)  #show help
      show_help;
      exit 0;
      ;;
  esac
done

shift $((OPTIND-1))  #This tells getopts to move on to the next argument.

# verify arguements have been set
if [ -z ${workflow_dir} ]; then
  echo "==> ERROR: Please specify the 'workflow_dir' arguement. Aborting script.";
  echo "==> Run '$0 -help' for additional info.";
  exit 1;
fi

#------------------------------------------------------------------------------
# VARIABLES
#------------------------------------------------------------------------------

SCRIPTPATH=`dirname "${BASH_SOURCE[0]}"`;
configs_dir="${SCRIPTPATH}/configs";
properties_dir="${SCRIPTPATH}/properties";
reports_dir="${SCRIPTPATH}/reports";
templates_dir="${SCRIPTPATH}/templates"
running=false;

#------------------------------------------------------------------------------
# MAIN
#------------------------------------------------------------------------------

# get environment info
color=$(grep -P "^color:" <(sudo cat /etc/salt/grains) | awk -F': ' '{ print $2 }');
if [ -z ${color} ]; then
  echo "==> ERROR: environment color can't be detected. Aborting script.";
  exit 1;
fi

# get hadoop cluster info
namenode=`hdfs getconf -confkey dfs.nameservices`
jobtracker=`hdfs getconf -confKey yarn.resourcemanager.cluster-id`

# get hdfs workflow list
for obj in `grep -P '^(?!(Found\s([0-9]*)\sitems))' <(hadoop fs -ls -R ${workflow_dir}) | awk '{ print $NF }' | sort`; do
  if is_workflow ${obj}; then

    # notify
    echo "==> Workflow file found: ${obj}"

    # set timestamps
    month=`date "+%m"`;
    day=`date "+%d"`;
    year=`date "+%Y"`;

    # reverse vars from obj path
    IFS='/' read -r -a path <<< "${obj}"
    if [ ${#path[@]} -ge 6 ]; then
      if [[ "${path[${#path[@]}-6]}" == "mdr" && "${path[${#path[@]}-5]}" == "standard" ]]; then
        provider="${path[${#path[@]}-7]}";
        database="${path[${#path[@]}-6]}";
        classification="${path[${#path[@]}-5]}";
        catalog="${path[${#path[@]}-4]}";
        schema="${path[${#path[@]}-3]}";
        table="${path[${#path[@]}-2]}";
      else
        provider="${path[${#path[@]}-6]}";
        database="${path[${#path[@]}-5]}";
        classification="all";
        catalog="${path[${#path[@]}-4]}";
        schema="${path[${#path[@]}-3]}";
        table="${path[${#path[@]}-2]}";
      fi
    else
      echo "==> ERROR: ${obj} needs to be a more complete path. Aborting script.";
      echo "==> Run '$0 -help' for additional info.";
      exit 1;
    fi

    # set start point if table name maches ${first_table} or if
    # first_table is undefined
    if [ ${first_table} ]; then
      if [ "${table}" == "${first_table}" ]; then
        running=true;
      fi
    else
      running=true;
    fi

    # process table
    if [ "${running}" = true ]; then

      # get config vars
      if [ "${classification}" != "all" ]; then
        config="${configs_dir}/${provider}_${database}_${classification}_${catalog}.txt"
      else
        config="${configs_dir}/${provider}_${database}_${catalog}.txt"
      fi
      if [ -f ${config} ]; then
        sql_user=`grep "sql_user" ${config} | awk -F' ' '{ print $2}'`
        sql_pass=`grep "sql_pass" ${config} | awk -F' ' '{ print $2}'`
        conn_string=`grep "conn_string" ${config} | awk -F' ' '{ print $2}'`
      else
        echo "==> ERROR: missing ${config} file. Aborting script.";
        exit 1;
      fi

      # construct job.properties output
      job_properties_dir="${properties_dir}/${provider}/${database}/${classification}/${catalog}/${schema}/${table}"
      job_properties_file="${job_properties_dir}/${table}_${year}${month}${day}_job.properties"

      # create job.properties file if missing
      if [ ! -f "${job_properties_file}" ]; then
        echo "==> Creating ${job_properties_file}";
        mkdir -p "${job_properties_dir}"
        if [ "${classification}" != "all" ]; then
          cp "${templates_dir}/job.properties" "${job_properties_file}";
        else
          cp "${templates_dir}/job_sans_class.properties" "${job_properties_file}";
        fi 

        # update values
        sed -i "s/_namenode/${namenode}/g" ${job_properties_file}
        sed -i "s/_jobtracker/${jobtracker}/g" ${job_properties_file}
        sed -i "s/_month/${month}/g" ${job_properties_file}
        sed -i "s/_day/${day}/g" ${job_properties_file}
        sed -i "s/_year/${year}/g" ${job_properties_file}
        sed -i "s/_provider/${provider}/g" ${job_properties_file}
        sed -i "s/_database/${database}/g" ${job_properties_file}
        sed -i "s/_classification/${classification}/g" ${job_properties_file}
        sed -i "s/_catalog/${catalog}/g" ${job_properties_file}
        sed -i "s/_schema/${schema}/g" ${job_properties_file}
        sed -i "s/_table/${table}/g" ${job_properties_file}
        sed -i "s/_sql_user/${sql_user}/g" ${job_properties_file}
        sed -i "s/_sql_pass/${sql_pass}/g" ${job_properties_file}
        sed -i "s/_conn_string/${conn_string}/g" ${job_properties_file}
        sed -i "s/_color/${color}/g" ${job_properties_file}
      fi

      # submit workflow to oozie
      job_output=$(oozie job -run -doas hdfs -oozie http://oozie.service.${color}.consul:11000/oozie -config ${job_properties_file});
      job_id=$( echo "${job_output}" | awk '{ print $2 }' );

      # verify job was submitted
      if [ -z "${job_id}" ]; then
        echo "==> ERROR: Job was not sucessfully submitted. Aborting script.";
        exit 1;
      else
        echo "==> Submitted Oozie job for ${provider}_${database}_${classification}_${catalog}_${schema}_${table}: ${job_id}";
      fi

      # wait for job to complete
      while grep -q -P "Status\s+:\s+RUNNING" <(oozie job -info ${job_id} -oozie http://oozie.service.${color}.consul:11000/oozie); do
        sleep 30;
      done

      # query oozie for job metrics
      mapfile -t metrics < <(grep -P "Status\s+:\s+(\w+)|Started\s+\:\s+([a-zA-Z0-9:-\s]+)|Ended\s+\:\s+([a-zA-Z0-9:-\s]+)" <(oozie job -info ${job_id} -oozie http://oozie.service.${color}.consul:11000/oozie))
      status=$(echo "${metrics[0]}" | awk -F': ' '{ print $2 }')
      start_time=$(echo "${metrics[1]}" | awk -F': ' '{ print $2 }')
      end_time=$(echo "${metrics[2]}" | awk -F': ' '{ print $2 }')

      # create report file
      report="${reports_dir}/${provider}_${database}_${classification}_${catalog}_${year}${month}${day}.csv"
      if [ ! -f ${report} ]; then
        if [ -f "${templates_dir}/report_default.csv" ]; then
          cp "${templates_dir}/report_default.csv" ${report};
        else
          echo "==> ERROR: Template not found ${templates_dir}/report_default.csv. Aborting script.";
          exit 1;
        fi
      fi

      # report metrics
      echo "${provider},${database},${classification},${catalog},${schema},${table},${status},${start_time},${end_time},${job_id}" >> ${report}

    fi

    # set exit if table name maches ${last_table}
    if [ ${last_table} ]; then
      if [ "${table}" == "${last_table}" ]; then
        exit 0;
      fi
    fi

  fi
done
